### PR TITLE
Add `parameter_frequencies` to operations.

### PR DESCRIPTION
**Pull request by dwierichs**
This is the first PR in a series to introduce the general parameter-shift functionalities to `qml.gradients.param_shift`.

Here we only add the operation property `parameter_frequencies` as well as the specific frequencies to all operations where applicable.

### CODE REVIEW

**Comment by josh146** [comment]
[sc-13053]

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -26,6 +37,11 @@
 
 <h3>Bug fixes</h3>
 
+* The operation `OrbitalRotation` previously was wrongfully registered to satisfy
+  the four-term parameter shift rule, it now will be decomposed instead when
+  using the parameter-shift rule.
+  [(#2180)](https://github.com/PennyLaneAI/pennylane/pull/2180)
```
Comment (josh146) [inline doc/releases/changelog-dev.md:43]: Quick question, rather than forcing the `OrbitalRotation` gate into a decomposition, could we just do?

```python
@property
def grad_recipe(self):
    coeffs, shifts = qml.gradients.generate_shift_rule(self.parameter_frequencies[0])
    return np.stack([coeffs, np.ones_like(coeffs), shifts]).T
```

Comment (josh146) [reply doc/releases/changelog-dev.md:43]: Basically, just a temporary hotfix to have the `OrbitalRotation` use the 8-term shift rule directly?

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -4,6 +4,17 @@
...
+* Parametric operations now have the `parameter_frequencies`
+  method that returns the frequencies with which a parameter
+  enters a circuit when using the operation.
+  [(#2180)](https://github.com/PennyLaneAI/pennylane/pull/2180)
+
+  The frequencies can be used for circuit analysis, optimization
+  via the `RotosolveOptimizer` and differentiation with the
+  parameter-shift rule. They assume that the circuit returns
+  expectation values or probabilities, for a variance
+  measurement the frequencies will differ.
```
Comment (josh146) [inline doc/releases/changelog-dev.md:16]: Still feels like this should be part of the name somehow, but I can never think of anything better than `parameter_frequencies` 😆 I try to have a 'zero or one underscore' rule when naming

File: pennylane/operation.py [hunk 3]
```diff
@@ -709,6 +718,52 @@ def generator(self):
         """
         return [None, 1]
 
+    def parameter_frequencies(self):
+        r"""Returns the frequencies for each operator parameter with respect
+        to an expectation value of the form
+        :math:`\langle \psi | U(\mathbf{p})^\dagger \hat{O} U(\mathbf{p})|\psi\rangle`.
+        These frequencies encode the behaviour of the operator :math:`U(\mathbf{p})`
+        on the value of the expectation value as the parameters are modified.
+        For more details, please see the :mod:`.pennylane.fourier` module.
```
Comment (josh146) [inline pennylane/operation.py:727]:
```suggestion
        :math:`\langle \psi | U(\mathbf{p})^\dagger \hat{O} U(\mathbf{p})|\psi\rangle`.

        These frequencies encode the behaviour of the operator :math:`U(\mathbf{p})`
        on the value of the expectation value as the parameters are modified.
        For more details, please see the :mod:`.pennylane.fourier` module.
```

File: pennylane/operation.py [hunk 4]
```diff
@@ -709,6 +718,52 @@ def generator(self):
...
+
+        Returns:
+            list[tuple[int or float]]: Tuple of frequencies for each parameter.
+                Note that only non-negative frequency values are returned.
```
Comment (josh146) [inline pennylane/operation.py:731]:
```suggestion
            Note that only non-negative frequency values are returned.
```

File: pennylane/operation.py [hunk 5]
```diff
@@ -709,6 +718,52 @@ def generator(self):
...
+
+        **Example**
```
Comment (josh146) [inline pennylane/operation.py:733]:
```suggestion
        **Example**

```

File: pennylane/operation.py [hunk 6]
```diff
@@ -709,6 +718,52 @@ def generator(self):
...
+        >>> op.parameter_frequencies()
+        [(0.5, 1), (0.5, 1), (0.5, 1)]
+        For operators that define a generator, the parameter frequencies are directly
+        related to the eigenvalues of the generator:
+        >>> op = qml.ControlledPhaseShift(0.1, wires=[0, 1])
+        >>> op.parameter_frequencies()
+        [(1,)]
+        >>> gen_eigvals = tuple(np.linalg.eigvals(op.generator[0] * op.generator[1]))
+        >>> qml.gradients.eigvals_to_frequencies(gen_eigvals)
+        (tensor(1., requires_grad=True),)
```
Comment (josh146) [inline pennylane/operation.py:744]:
```suggestion
        [(0.5, 1), (0.5, 1), (0.5, 1)]

        For operators that define a generator, the parameter frequencies are directly
        related to the eigenvalues of the generator:

        >>> op = qml.ControlledPhaseShift(0.1, wires=[0, 1])
        >>> op.parameter_frequencies()
        [(1,)]
        >>> gen_eigvals = tuple(np.linalg.eigvals(op.generator[0] * op.generator[1]))
        >>> qml.gradients.eigvals_to_frequencies(gen_eigvals)
        (tensor(1., requires_grad=True),)
```

Line breaks are required for this to render properly :)

https://pennylane--2180.org.readthedocs.build/en/2180/code/api/pennylane.operation.Operation.html#pennylane.operation.Operation.parameter_frequencies

File: pennylane/operation.py [hunk 7]
```diff
@@ -709,6 +718,52 @@ def generator(self):
...
+            if op is None:
+                raise OperatorPropertyUndefined(
+                    f"Operation {self.name} does not have parameter frequencies."
+                )
+
+            if not isinstance(op, np.ndarray):
+                op = op.matrix
+            gen_eigvals = tuple(np.linalg.eigvals(op))
+            coeff = np.abs(coeff)
+            return tuple(coeff * val for val in qml.gradients.eigvals_to_frequencies(gen_eigvals))
```
Comment (josh146) [inline pennylane/operation.py:761]: Would be very happy if you want to leave this if statement branch as a TODO, and simply raise `NotImplementedError` for now!

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (josh146) [reply doc/releases/changelog-dev.md:43]: Funnily enough, the [known decomposition](https://github.com/PennyLaneAI/pennylane/blob/d09598c6e91424569df007ea51f0a097d806c40f/pennylane/ops/qubit/qchem_ops.py#L636-L652) has 4 parametrized gates, so there is no savings re: number of quantum executions

File: pennylane/ops/qubit/parametric_ops.py [hunk 8]
```diff
@@ -56,6 +56,9 @@ class RX(Operation):
     grad_method = "A"
     generator = [PauliX, -1 / 2]
 
+    def parameter_frequencies(self):
+        return [(1,)]
```
Comment (josh146) [inline pennylane/ops/qubit/parametric_ops.py:60]: Very very minor, but if a the operator has parameter frequencies that don't depend on `self`, we could ovewrite as a class attribute:
```suggestion
    parameter_frequencies = [(1,)]
```

the main (minor) advantage being that you can query an uninstantiated class

File: pennylane/ops/qubit/parametric_ops.py [hunk 9]
```diff
@@ -826,6 +850,14 @@ class CRX(Operation):
     def label(self, decimals=None, base_label=None):
         return super().label(decimals=decimals, base_label=base_label or "RX")
 
+    def parameter_frequencies(self):
+        return [
+            (
+                0.5,
+                1,
+            )
+        ]
```
Comment (josh146) [inline pennylane/ops/qubit/parametric_ops.py:859]: black didn't like this? 😱

File: pennylane/ops/qubit/qchem_ops.py [hunk 10]
```diff
@@ -114,23 +122,23 @@ def label(self, decimals=None, base_label=None):
         return super().label(decimals=decimals, base_label=base_label or "G")
 
 
-class SingleExcitationMinus(Operation):
-    r"""SingleExcitationMinus(phi, wires)
-    Single excitation rotation with negative phase-shift outside the rotation subspace.
+class SingleExcitationPlus(Operation):
+    r"""SingleExcitationPlus(phi, wires)
```
Comment (josh146) [inline pennylane/ops/qubit/qchem_ops.py:126]: Did the order of the operations swap?

File: pennylane/ops/qubit/qchem_ops.py [hunk 11]
```diff
@@ -588,6 +617,16 @@ class OrbitalRotation(Operation):
...
 
+    def parameter_frequencies(self):
+        return [
+            (
+                0.5,
+                1.0,
+                1.5,
+                2.0,
+            )
+        ]
```
Comment (josh146) [inline pennylane/ops/qubit/qchem_ops.py:628]: I think if you rewrite this (and above) as one line, black shouldn't complain?

File: pennylane/ops/qubit/qchem_ops.py [hunk 12]
```diff
@@ -535,8 +563,10 @@ class OrbitalRotation(Operation):
 
     * Number of wires: 4
     * Number of parameters: 1
-    * Gradient recipe: The ``OrbitalRotation`` operator satisfies the four-term parameter-shift rule
-      (see Appendix F, https://arxiv.org/abs/2104.05695)
+    * Gradient recipe: The ``OrbitalRotation`` operator does *not* satisfy the
+      four-term parameter-shift rule but a higher-order general parameter-shift rule
+      (with 4 equidistant frequencise :math:`\{0.5, 1, 1.5, 2\}` see,
+      https://arxiv.org/abs/2107.12390)
```
Comment (josh146) [inline pennylane/ops/qubit/qchem_ops.py:569]: Really nice catch! Although I might simply say
```suggestion
    * Gradient recipe: The ``OrbitalRotation`` operator has 4 equidistant frequencies
      :math:`\{0.5, 1, 1.5, 2\}`, and thus permits an 8-term parameter-shift rule.
      (see https://arxiv.org/abs/2107.12390).
```

Comment (josh146) [reply pennylane/ops/qubit/qchem_ops.py:569]: That is, saying it does not support a 4-term rule only really makes sense if you know we used to say it does 😆

File: pennylane/ops/qubit/qchem_ops.py [hunk 13]
```diff
@@ -562,8 +592,7 @@ class OrbitalRotation(Operation):
     """
     num_wires = 4
     num_params = 1
-    grad_method = "A"
-    grad_recipe = four_term_grad_recipe
+    grad_method = "F"
```
Comment (josh146) [inline pennylane/ops/qubit/qchem_ops.py:595]:
```suggestion
    grad_method = None
```
this should force the decomposition

**Review by josh146 — COMMENTED** [review]
Thanks @dwierichs! Have left only minor comments :)

File: pennylane/ops/qubit/parametric_ops.py (hunk 8 above)
Comment (josh146) [reply pennylane/ops/qubit/parametric_ops.py:60]: wait never mind, I thought `parameter_frequencies` was a property, but it is a method - ignore the above comment!

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (dwierichs) [reply doc/releases/changelog-dev.md:43]: Ah, yes, that definitely is a better option :)
I'd say the standard case for gates is that the number of decomposition shifts and generalized shifts exactly matches. :)

File: doc/releases/changelog-dev.md (hunk 2 above)
Comment (dwierichs) [reply doc/releases/changelog-dev.md:16]: Yeah, I was thinking about this as well, but I'd say it is okay, because the whole fourier module and Rotosolve also assume this context of expectation value/probs outputs :)

File: pennylane/operation.py (hunk 7 above)
Comment (dwierichs) [reply pennylane/operation.py:761]: Is it not okay like this? I found it nice that this allows to already understand the logical flow for default operations that do not define custom `parameter_frequencies`.

File: pennylane/ops/qubit/parametric_ops.py (hunk 8 above)
Comment (dwierichs) [reply pennylane/ops/qubit/parametric_ops.py:60]: Should it be a property? :thinking:

File: pennylane/ops/qubit/qchem_ops.py (hunk 10 above)
Comment (dwierichs) [reply pennylane/ops/qubit/qchem_ops.py:126]: Yes, exactly, in order to make it `..Plus` first, matching the `DoubleExcitation` order. Sorry for being pedantic :sweat_smile:

File: pennylane/ops/qubit/qchem_ops.py (hunk 11 above)
Comment (dwierichs) [reply pennylane/ops/qubit/qchem_ops.py:628]: I think black _did_ this... :thinking: Will look into it.

File: pennylane/ops/qubit/qchem_ops.py (hunk 13 above)
Comment (dwierichs) [reply pennylane/ops/qubit/qchem_ops.py:595]: Ah yes of course! 
I'll skip this one and use the eight-term rule above instead?

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (josh146) [reply doc/releases/changelog-dev.md:43]: Yeah, I was just hoping that maybe the known decomposition was poor 😆

File: pennylane/operation.py (hunk 6 above)
Comment (dwierichs) [reply pennylane/operation.py]: Thanks, nice catch!

File: pennylane/operation.py (hunk 7 above)
Comment (josh146) [reply pennylane/operation.py:761]: Oh, this was more a comment that the generator will soon be breaking :) But definitely fine as is, no action needed here

File: pennylane/ops/qubit/qchem_ops.py (hunk 11 above)
Comment (dwierichs) [reply pennylane/ops/qubit/qchem_ops.py]: Ah, if one leaves a trailing coma although the tuple has more than one entry, black will do this for some reason :)

File: pennylane/operation.py (hunk 7 above)
Comment (dwierichs) [reply pennylane/operation.py:765]: Ah, I see :) Yeah, I went for postponing any such trouble to later parts of this series of PR, just to get started :)

**Review by josh146 — APPROVED** [review]
Looks good to me @dwierichs!

File: pennylane/ops/qubit/qchem_ops.py [hunk 14]
```diff
@@ -587,12 +594,19 @@ class OrbitalRotation(Operation):
         ),
         -1 / 2,
     ]
+    parameter_frequencies = [(0.5, 1.0, 1.5, 2.0)]
+
+    @property
+    def grad_recipe(self):
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/ops/qubit/qchem_ops.py:600]: @trbromley yes, this one was super odd - I assumed codefactor was incorrect/buggy, since we are simply overwriting a property of the parent class?

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/ops/qubit/qchem_ops.py:600]: https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/operation.py#L626-L640

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/ops/qubit/qchem_ops.py:600]: I'll make a PR to fix this :)
